### DISCUSSION TITLE
Continuous live stream on dashboard and camera groups

### DISCUSSION THREAD

**Discussion by kensand** [blakeblackshear/frigate#12697] [Ideas]
### Please upvote this if you would like to see this feature added! The code is already written!
This is important so maintainers can gauge support for the feature.

If you have another idea for how this can be implemented, perhaps in a less easily misconfigured manner, please comment below. One idea I had was to enable continuous live streams on a per-camera basis.

-----

Per #12684 and #12682, continuous live streams of cameras on the dashboard and camera groups is not directly supported by Frigate. Cameras only begin streaming if Automatic Live View is enabled in the settings page and once motion is detected. If Automatic Live View is disabled, snapshot views of the cameras on the dashboard and camera groups will only update once per minute.

For my personal use case, I have both the bandwidth and server resources to support a continuous live stream for all my cameras, so I prefer to be able to view these live streams continuously regardless of whether there is ongoing motion. 

The suggested workarounds for this issue are:
1. Use [birdseye](https://docs.frigate.video/configuration/birdseye) in continuous mode. However, this is not the same quality as individual cameras and cannot be rearranged/resized according to user preferences. Additionally, this does not enable users to enlarge a camera of interest simply by clicking on it. It also spins up another ffmpeg process, albeit one that should be rather efficient.
2. Reduce motion threshold on the live streaming cameras to 1, causing them to always be active. This has the desired effect, but does not allow users to reasonably configure objection detection as all detectors are always active, further using up resources.

Some other downsides to not having this feature:
1. Cameras with motion detection disabled cannot be live streamed at all, even with the Automatic Live View enabled. The only way to do it would be to use birdseye mode, which has its own downsides listed above.
2. Motion under the motion detection threshold will not cause the cameras to be made active.
3. Motion detection may be masked for the camera, but a user may still want to be able to monitor it in realtime; for example, the wind on a flag or traffic on a roadway.
4. A camera may have a hardware issue that causes it to continue responding to stream requests but not to update the image, which would not be readily apparent to the user.

Frigate maintainers have expressed concerns about the possibility of misconfiguration with this setting, which would cause a greater use of system resources and bandwidth. While I can't say I agree with this opinion, I understand it. I will note that I personally spent over an hour trying to figure out why live streams weren't working, and that I had to dive into the frontend code to understand the problem. At a minimum, I think this needs to be added to the FAQ and/or live stream page in the docs (which I plan to make a separate PR for).

In the meantime, I will be maintaining [my own Frigate fork](https://github.com/kensand/frigate) containing #12684, if anyone else needs this feature. It should automatically sync with the upstream repo using the [Github pull app](https://github.com/apps/pull)

**Comment by danielpmorris87** [comment]
This would be an awesome feature to have. Id love to use the Frigate UI when I am working/gaming but I am forced to use Home Assistant for continuous viewing.

**Comment by nicoleise** [comment]
If your premise is correct, then I would like this feature also. I thought Frigate already had a fallback behavior for cameras where motion isn't enabled, since the live mechanism seems to depend on it.

Your second list suggests that the most relevant way to implement this would be on a per-camera basis. 

So, under config under each camera an optional entry like `always-live-on-dashboard: True`. If true, and if no motion is configured, or if motion is configured but not currently detected, the camera would display a live feed but without the accenting frame and icons associated with motion. Those highlights could still be shown if motion is detected, if for example this option is used on a motion enabled camera to ensure that the view is live even if the motion of interest isn't detected. 


To provide a specific use case example: We will have some cameras that aren't used for automations and are situated in areas you can't get to or from without being picked up by several other motion enabled cameras. Some of these will overlook our backyard, the playground-ish area for the kids and our fields in hilly terrain. In other words, motion would likely frequently be triggered by stuff we don't care about, like plants moving, tractors in the field, shadows from clouds, etc. It would likely not be triggered by what we actually want, for example the kids in the playground, because of the distance to those events.

We cannot set up motion in a way that would work. For example, a large tractor in the field in the foreground is of no interest, while a small camouflaged hunter 1 km away is of high interest. We don't care about a person on the ride on lawn mower, but we'd like to be able to see that our kids are ok in their playground area if we're inside cooking and they're outside playing. But that's a partially obstructed view at around 70 m distance. We don't care about the horses themselves, but if the fence is suddenly drooping, we'd be very interested. 

So we'd like to simply be able to see that camera live (together with other cameras) under certain circumstances, but we wouldn't want/need (mostly irrelevant) motion events to clog up events. 

I thought non motion cameras in Frigate would already exhibit this behavior om dashboards, so I can only support the suggestion.

**Reply by kensand** [reply]
> I thought Frigate already had a fallback behavior for cameras where motion isn't enabled, since the live mechanism seems to depend on it.

Unfortunately no; if motion isn't enabled, the image on the dashboard and camera groups only update once every minute, regardless of the AutoLive setting or any configuration.

> Your second list suggests that the most relevant way to implement this would be on a per-camera basis.
>
> So, under config under each camera an optional entry like always-live-on-dashboard: True. If true, and if no motion is configured, or if motion is configured but not currently detected, the camera would display a live feed but without the accenting frame and icons associated with motion. Those highlights could still be shown if motion is detected, if for example this option is used on a motion enabled camera to ensure that the view is live even if the motion of interest isn't detected.

I'm open to doing so; right now the PR I have open (and my own fork) takes a simpler route though, and just changes the AutoLive toggle into a dropdown allowing users to choose between three settings:

* Auto mode (default) - begin streaming when activity (motion) is detected.
* Static mode - update camera images on the live dashboard once per minute.
* Continuous mode - stream cameras regardless of activity.

This mode affects all cameras visible to the user and regardless of camera group or dashboard.

Adding an additional configuration on a per-camera basis as you describe would definitely add some work and complexity, but I would be happy to update my PR if the maintainers prefer it.

-----

Thanks for your feedback and additional use case. Hopefully the maintainers will take it into account :smiley:

**Reply by NickM-27** [reply]
I think it's important to define what a non motion camera is. There is no requirement for object detection or alerts to be setup. Motion detection is enabled by default so unless you are explicitly disabling motion detection (different from object detection) on the camera then the live view will work the same as any other case.

**Comment by dopeytree** [comment]
Is birdseye mode still a thing in 0.14?

**Reply by NickM-27** [reply]
Yes

**Reply by roycamp** [reply]
UX issue

**Comment by pschroee** [comment]
I really like this feature. I want to setup a dedicated pc with frigate installed. The pc should auto login and open frigate in firefox in kiosk mode. I want to see the live view of the cameras all times. The pc, on which frigate is installed, has the resources to always show the livestreams of the cameras.
For the concerns of misconfiguration, you can add a warning label to the setting or you have to set config entry (e.g. ui -> enable continuous live view).

**Comment by conorlap** [comment]
+1 for this, would be very useful to have ability to configure this!

**Comment by saket424** [comment]
@kensand 
I would like the complete opposite of the problem you are addressing and so having the ability to configure whether we want a camera stream to render all the time or the stream to render just on demand would be excellent. Assume you have 30 cameras and all you care is on demand live view and no recording or object detection is necessary. Currently my measurements indicate that frigate subscribes to all 30 streams and pulls a copy of the feeds even when there is no one actively viewing the feeds. Go2rtc is frugal with bandwidth and only pulls when someone requests to view or analyze the feeds. I would like that frugality to also carryover to frigate. I hope I articulated my point.

**Reply by NickM-27** [reply]
Frigate always decodes the streams for a number of features that require it. Using frigate without detection and without recording is a very uncommon use case for Frigate

**Comment by bill-web7** [comment]
I also spent hours debugging my frigate installation because it isn't live and honestly I couldn't conceive this is the intended behavior. As far as I'm concerned this is a duh feature and live should be the default. The maintainers arguments sound like they are worried about people using a Tamagotchi to run 20 cameras. I really appreciate the work they are doing but this is 2024. In my 30 years as a C/C++ programmer and CTO I always fought against the "we know best" mindset on display here. The most obvious first test people will perform is to check if they are at least getting live video. I wonder how many people gave up on frigate and HA because of this frustration when that simple expectation was not met.
So it seems that my only solution is to use frigate for recording and HA camera for live feeds. Seems like a bigger waste of resources than the proposed solution.

**Reply by hua-bing** [reply]
> I always fought against the "we know best" mindset on display here.

Somewhat agree.  The best example would be the trend of using relative time in almost anywhere.  I understand that it might be (questionable) good for social platform, it is not a good idea to make it default and even not configurable for version control systems like Github and Bitbucket.  A little bit insane to me.

**Reply by NickM-27** [reply]
You are greatly underestimating how many resources restreaming / high res live view uses. During the beta testing for this version there was an issue where even if a camera wasn't shown on the screen (it was below the fold) it was still live and users saw very high cpu usage from their 20 cameras that were seeing motion. 

It's possible once frigate supports multiple assignable live streams so it can always show the sub stream on the grid view and the higher res stream on the single cam view this could be more approachable. 

But again, I'm not understanding why you want to see a live view of still image when an actual image is exactly the same thing. During the 0.14 beta there were some questions on this but after the explanation users said it made sense and said they preferred it especially when mobile to save bandwidth. 

It's always possible things can change, and we have some ideas to implement something like this. As someone who has provided support to frigate users for 3 years it is clear this will be an issue for many if enabled with unrealistic expectations, and we will have to fill that support burden

**Reply by bill-web7** [reply]
Let me give you my example and maybe that will help me understand. I am using up to date HAOS on a i7 16 gb 1080 Coral USB system on a gig wired network. I have each camera (10) include the time on each frame. The motion detection doesn't pick up the seconds changing. Maybe there is a setting I'm not tweaking. If there is please point me to the right place. I just feel it is expected behavior that any change should be reflected in the video stream.

**Reply by NickM-27** [reply]
it is typically recommended to put a motion mask on the timestamp so it does not trigger motion detection. If there is no motion mask and no motion detected then thetimestamp might just be a bit small.

The main question is why do you want it to be live? If the only changing in the entire image is the timestamp, why waste bandwidth and CPU resources on that when a still image is the same thing minus the time updating. In general it sounds like in your case the desire is for it to be live even if there is no practical reason for it to be live when Frigate immediately switches to the live view when motion is detected.

**Reply by bill-web7** [reply]
How small is small and who gets to decide?

**Reply by NickM-27** [reply]
You can always increase the sensitive of motion detection in the motion tuner

**Reply by kensand** [reply]
The streams are still being processed anyways, so it should not affect server CPU (and it hasn't in my experience). To be clear, my fork contains changes _only_ on the frontend, so the only effect should be bandwidth consumption. 

At this point, it seems clear that feature is desirable for many people. How many people need to upvote this before you consider merging it @NickM-27? It can be gated behind settings for safety to minimize support burden. Frigate users aren't stupid - a non-default setting with an adequate warning of bandwidth usage should be sufficient...

**Reply by NickM-27** [reply]
> The streams are still being processed anyways, so it should not affect server CPU (and it hasn't in my experience). To be clear, my fork contains changes only on the frontend, so the only effect should be bandwidth consumption.

that is incorrect, the MSE / WebRTC restream requires CPU from go2rtc as well, and can be quite high especially for 4k cameras

> At this point, it seems clear that feature is desirable for many people. How many people need to upvote this before you consider merging it @NickM-27? 

Like I said, I think this will be more approachable when multi stream selection is implemented so users can configure a sub stream to be used on the dashboard if they wish which will allow full res stream viewing on the main camera view without using a lot of CPU resources on the dashboard if continuous is enabled

**Reply by kensand** [reply]
So in other words, you don't want to address this request from your users, despite your previous statements on the PR and Issue. That's disappointing 😞 .

**Reply by NickM-27** [reply]
I said a separate feature needs to be implemented first before this can be considered, nowhere did I say this would not be addressed. Also, I am not making a decision only stating my personal opinion

**Reply by kensand** [reply]
> I said a separate feature needs to be implemented first before this can be considered, nowhere did I say this would not be addressed.

So, as a maintainer, you are making the _decision_ not to merge the existing implementation I provided. My implementation works and is relatively tidy, and I am happy to hide the functionality behind additional settings and warnings, but the _decision_ not to merge is squarely within your purview. The requirement that another feature (multistream) be implemented first is effectively moving goalposts.

> Also, I am not making a decision only stating my personal opinion

You may feel you are voicing your opinion, but by default, a maintainer's opinion is a decision, unless overridden by other maintainers.


It's disappointing, but at least I am free to maintain my fork. I truly would love to be proven wrong about your _decision_ though :smile:

**Reply by hawkeye217** [reply]
As another maintainer of Frigate, I'll jump in here to say that we certainly see the interest in this feature and we have discussed how to best implement something like it. As I mentioned on your original PR, a more granular approach to this feature is preferable so that users have the ability to stream none, some, or all of their cameras continuously rather than none or all.

As we've mentioned several times, a one-size-fits-all approach is not feasible because of the support issues that it would lead to. Many users are unaware of the server and client requirements for continuous streaming, and as volunteers to the project, we don't want to (nor do we have the time to) field dozens of support requests for users who would enable continuous streaming and then say "my 5 year old tablet doesn't work with Frigate, it's a bug". Many users run Frigate on under-powered SBCs or older processors, so additional documentation will be necessary to help users understand server and client requirements. I hope you can see and appreciate the support aspect of this feature.

Additionally, as @NickM-27 mentioned, we'd like to provide users with granular control and the ability to choose any stream from their cameras for the live option. A continuous streaming feature makes more sense in that context.

I've begun implementing this feature already and hope that it can be a part of Frigate 0.16. When it's merged in dev, feel free to test and give your feedback.

**Reply by kensand** [reply]
> As another maintainer of Frigate, I'll jump in here to say that we certainly see the interest in this feature and we have discussed how to best implement something like it.

I appreciate you jumping in. Is there somewhere those discussions are visible, or are they made via private channels? I would like to know what was proposed and would be acceptable to the maintainers as a whole, if possible.

> As I mentioned on your original PR, a more granular approach to this feature is preferable so that users have the ability to stream none, some, or all of their cameras continuously rather than none or all.

I would be happy to implement a granular approach, if I had reason to believe that it would have an additional likelihood of being merged, but it seemed your concern over supporting foot-shooting users trumped that approach as well; wouldn't it be just as likely for someone to set a granular config as what I had in the original implementation?

> As we've mentioned several times, a one-size-fits-all approach is not feasible because of the support issues that it would lead to.

Please clarify what you mean by "one-size-fits-all". Are you referring to this not being granular on a per-camera basis? Or that providing continuous live streams on the dashboard/camera groups are not applicable to all users? or something else?

> Many users are unaware of the server and client requirements for continuous streaming, and as volunteers to the project, we don't want to (nor do we have the time to) field dozens of support requests for users who would enable continuous streaming and then say "my 5 year old tablet doesn't work with Frigate, it's a bug". Many users run Frigate on under-powered SBCs or older processors, so additional documentation will be necessary to help users understand server and client requirements. I hope you can see and appreciate the support aspect of this feature.

As I have said before, I understand your concern, but I don't agree with it. I believe the _lack_ of this feature is likely causing more confusion for users than including it with gratuitous warnings would, as @bill-web7 and others in this discussion have noted.

> Additionally, as @NickM-27 mentioned, we'd like to provide users with granular control and the ability to choose any stream from their cameras for the live option. A continuous streaming feature makes more sense in that context.

> I've begun implementing this feature already and hope that it can be a part of Frigate 0.16. When it's merged in dev, feel free to test and give your feedback.

When you write "I've begun implementing this feature", are you referring to the multi-stream feature or continuous live stream feature, or both?

I agree continuous live streaming does make more sense in the context of having multiple streams - I use sub-streams via separate camera configurations for recording vs streaming. However, the lack of live streaming is still an issue for me regardless of the multi-stream implementation, and I am trying to gauge whether I should embark on writing a PR for granular, per-camera-enabled continuous live stream in the meantime.



Thank you for the additional background though, it is appreciated.

**Reply by bill-web7** [reply]
Following Akin's First Law I provide a few simple numbers from my setup. Following the suggestion of NickM I tried different values of:
motion:
  threshold: 
and recorded the effect on CPU use and Network throughput out. Each of the 10 cameras was setup to use the low res stream. I included only cameras where there is very little movement at this time of day. I setup 2 consumers, one wired windows system and one iphone on wifi. I also made sure to turn off everything I could think of on the HA PC. I waited a minute after restarting Frigate for things to settle.

Threshold          CPU %        Throughput out        Comment
30 (default?)      22-24         0.100 - 0.290 MB/s     Some cameras didn't update for at lease 30 seconds.
1 (min)               28-32         0.400 - 0.580 MB/s     This setting gave me what I wanted to see; each second updated.
100                    21-22         0.089 - 0.096 MB/s      Some cameras took long to startup.

Thanks to NickM for his suggestion it allows me to get on with my HA integration as in now have what I expected to see. I feel especially dumb not thinking of looking in the docs for a way to set the threshold. Especially since I have written at least 100,000 lines of C++ code using OpenCV.

**Comment by hua-bing** [comment]
I like this feature.  And I would vote for making it as the default.  Personally I thought there are issues with my cameras and/or setups when I saw some cameras did not update and spent a lot of time to only find out that it is an expected behavior.
I am also not sure if birdview needs to exist if there is live stream from the snapshot.  It might be a tradition to have a birdview for a surveillance system.  However, the birdview is a little bit hard to use (and redundant when live stream is available from snapshot view).  I often find myself clicking on birdview cameras trying to get to the camera view.

**Reply by NickM-27** [reply]
Many use birdseye view for showing on a TV or tablet that isn't powerful enough to display all the different streams in the webui. Birdseye is not shown in the webui by default

**Reply by dedors** [reply]
I can totally relate! When I first started using CCTV and Frigate, I had no idea that the Live View only displays the actual live feed when motion is detected. I expected it to show all cameras streaming live, and I ended up wasting quite a bit of time before I figured that out.  I'm still stunned because I never thought about that possible behavior, because as I'm in the testing phase, I didn't setup the masks for all the cameras, as they all do have a temporary location. That includes the timestamps, so some cameras did work live while others didn't.

**Comment by roycamp** [comment]
Please fix

**Reply by hawkeye217** [reply]
https://github.com/blakeblackshear/frigate/discussions/12697#discussioncomment-11220168

**Comment by foxtrotcz** [comment]
I see that adding support is planned. I just want to say why I would like it.

I disabled motion detection because I want to have minimal cpu usage. Maybe I will use it in future but for now I wanted just basics, which means just video stream without any detection.

But this makes it impossible for me to see multiple cameras live when I want to check them.

Maintainers said that continously streaming cameras with no motion is wasting energy. I understand that point.
But in my case the motion detection itself is disabled which also saves energy on server but makes it impossible to view multiple live cameras. 
And the energy lost when I check the cameras few times a day seems small. 

I understand that maintainers know the how frigate should ideally be used. But please allow us new users this simple use cases even if they are not ideal. It's just something I would be comfortable with and get frigate going and then I can take my time with learning other better ways how to use frigate.

Thanks.

**Reply by hawkeye217** [reply]
Support has already been added for the next version of Frigate. 

https://github.com/blakeblackshear/frigate/pull/16447
